File: vo/req/bill.go
package req

import (
	"mime/multipart"
	"time"

	"github.com/Guo-Chenxu/pay-log/pkg/pageutil"
)

type UploadBillReq struct {
	Channel int8                  `form:"channel" binding:"required"`
	File    *multipart.FileHeader `form:"file"    binding:"required"`
}

type ListBillsReq struct {
	Year      int    `form:"year"`
	Month     int    `form:"month"`
	SortOrder string `form:"sort_order"` // asc | desc (by amount), empty = time desc
	pageutil.PageReq
}

type ListRangeBillsReq struct {
	StartYear  int    `form:"start_year"`
	StartMonth int    `form:"start_month"`
	EndYear    int    `form:"end_year"`
	EndMonth   int    `form:"end_month"`
	SortOrder  string `form:"sort_order"`
	pageutil.PageReq
}

type AddManualBillReq struct {
	Channel         int8      `json:"channel"          binding:"required"`
	TransactionTime time.Time `json:"transaction_time" binding:"required"`
	Counterparty    string    `json:"counterparty"`
	Description     string    `json:"description"`
	Category        string    `json:"category"`
	BillType        int8      `json:"bill_type"        binding:"required"`
	Amount          string    `json:"amount"           binding:"required"`
	PaymentMethod   string    `json:"payment_method"`
	Remark          string    `json:"remark"`
	IsInvestment    bool      `json:"is_investment"`
}

File: web/src/api/bill.ts
import type { BillType, Channel } from '../constants/bill'
import client from './client'

export interface ManualBillPayload {
  channel: Channel
  transaction_time: string
  counterparty: string
  description: string
  category: string
  bill_type: BillType
  amount: string
  payment_method: string
  remark: string
  is_investment: boolean
}

export const uploadBill = (file: File, channel: Channel) => {
  const form = new FormData()
  form.append('file', file)
  form.append('channel', String(channel))
  return client.post('/bill/upload', form)
}

export const addManual = (data: ManualBillPayload) => client.post('/bill/manual', data)

export const listBills = (year: number, month: number, page = 1, page_size = 20, sort_order = '') =>
  client.get('/bill/list', { params: { year, month, page, page_size, sort_order: sort_order || undefined } })

export const listRangeBills = (
  startYear: number, startMonth: number,
  endYear: number, endMonth: number,
  page = 1, page_size = 20, sort_order = ''
) =>
  client.get('/bill/list-range', {
    params: { start_year: startYear, start_month: startMonth, end_year: endYear, end_month: endMonth, page, page_size, sort_order: sort_order || undefined },
  })

export const deleteBill = (id: string) => client.delete(`/bill/${id}`)
